Add Multiple Language Support

littlecube8152 opened the pull request:
Add framework of supporting multiple languages and adds Python support implementation.

littlecube8152 committed:
Move C++ related make logic into langauge class

littlecube8152 committed:
Refactor compilation details into internal.compilation

Now all compilation depends on internal.compilation, including language
supports. In addition, makefile now makes all produced files into build
directory instead of littering all of them in the same dir.

littlecube8152 committed:
Add Python language support and fix minor bugs

This commit adds Python languages supports. In addition, some bugs
including wrong compilation result and incorrect model solution invoked
are fixed in this commit.

littlecube8152 committed:
Merge branch 'develop' into multi-lang

littlecube8152 committed:
Attempt to fix decode error

littlecube8152 committed:
Attempt to fix workflow part 2: make

littlecube8152 committed:
Attempt to fix workflow part 2B: makefile

littlecube8152 committed:
Fix type errors and apply ruff format

OmeletWithoutEgg committed:
Try to fix CI by changing macos version

OmeletWithoutEgg committed:
Disable timeout in CI

OmeletWithoutEgg committed:
Change workflow back and try gmake

OmeletWithoutEgg commented on internal/compilation/languages/abc.py:
Maybe we can name this base.py?

littlecube8152 replied:
Sounds good

littlecube8152 committed:
Try nonblocking IO for retrieving outputs

littlecube8152 committed:
Remove -Wshadow and -Wconversion from default flags

littlecube8152 committed:
Add compilation stress test

littlecube8152 committed:
Attempt to fix previous test

littlecube8152 committed:
Revert remove -Wshadow and -Wconversion flags

littlecube8152 committed:
Increase compilation time limit

littlecube8152 committed:
Remove compilation stress test and improve wait output implementation

Compilation stress test is removed since it fails to be a good test;
reading the humongous output simply takes too much time and hurts the disk.

littlecube8152 committed:
Rename abc -> base

littlecube8152 committed:
Fix aplusb-py and make it a test

littlecube8152 committed:
Remove incorrect checker assertion

littlecube8152 committed:
Update gitignore and include hash verification in unit test

littlecube8152 committed:
Delegates single file compilation to target Makefile

OmeletWithoutEgg committed:
Fix README.md link

littlecube8152 commented on internal/runner.py:
```diff
@@ -236,36 +236,45 @@ def wait_for_outputs(proc: Process) -> tuple[str, str]:
+    if proc.stderr is not None:
+        fd = proc.stderr.fileno()
+        flags = fcntl.fcntl(fd, fcntl.F_GETFL)
+        fcntl.fcntl(fd, fcntl.F_SETFL, flags | os.O_NONBLOCK)
```
I think not. The function ends when either the stream ends (therefore the pipes are useless), or the process terminates.

OmeletWithoutEgg replied:
I just found that there is `os.set_blocking(f.fileno(), False)`. Maybe we can use it so that we won't need to `import fcntl`?

littlecube8152 commented on internal/compilation/single.py:
```diff
@@ -0,0 +1,110 @@
+                    allout += stdout
+                    allerr += stderr
+                finally:
+                    if compile_process is not None:
```
I am not sure what happens when the `Process` constructor fails (throws an exception). You are right about the second one.

littlecube8152 commented on internal/compilation/makefile.py:
```diff
@@ -0,0 +1,149 @@
+                    stderr=subprocess.PIPE,
+                    time_limit_sec=compilation_time_limit_sec,
+                    memory_limit_mib=compilation_memory_limit_mib,
+                    env={
```
You are correct. Apparently I don't know Python dictionaries.

littlecube8152 committed:
Minor fixes and formatting

OmeletWithoutEgg commented on internal/compilation/languages/python.py:
```diff
@@ -0,0 +1,43 @@
+                self.context.path.script_dir,
+                "internal/compilation/languages/makefiles/python.target.Makefile",
+            ),
+            env={},
```
I guess it should be `env=os.environ` here.

littlecube8152 replied:
No, it serves as extra environment for the Makefile to work. I should rename this to extra_env and add a comment.

OmeletWithoutEgg commented on internal/formatting.py:
```diff
@@ -138,7 +138,7 @@ def format(color: str, content: str):
 
     def print_reason(self, reason: str):
-        # TODO: the following terminal width threshold should be configurable via global settings
+        # : the following terminal width threshold should be configurable via global settings
```
I'm not sure why `TODO` is deleted here.

OmeletWithoutEgg commented on internal/commands/gen.py:
```diff
@@ -216,6 +217,8 @@ def command_gen(
             with open(context.path.testcases_hashes, "r") as f:
                 official_testcase_hashes: dict[str, str] = json.load(f)
             formatter.print_hash_diff(official_testcase_hashes, testcase_hashes)
+            if official_testcase_hashes != testcase_hashes:
```
Maybe we can use `print_hash_diff_with_exit` here, just like `print_compile_string_with_exit`.

OmeletWithoutEgg commented on internal/compilation/makefile.py:
```diff
@@ -0,0 +1,142 @@
+from .languages import languages
+
+
+# if platform.system() == "Darwin":
```
Shall we ask user to ensure their make is `gmake`, or shall we auto-detect and choose here?

littlecube8152 committed:
Add Makefile template and slightly refactor compilation

littlecube8152 committed:
Minor fix formatting and code format

littlecube8152 committed:
Remove redundant check and exit

OmeletWithoutEgg reviewed (approved):
LGTM
